From ab956eadf5787366d3f48c89276b6644f16193fc Mon Sep 17 00:00:00 2001
From: Balamurugan Ramar <quic_bramar@quicinc.com>
Date: Sun, 4 Aug 2024 22:18:30 +0530
Subject: [PATCH] wifi: ath12k: remove the arvif lookup in ath12k_wmi_bcn_tmpl

Lock is missed to add in function ath12k_wmi_bcn_tmpl while
accessing the critical data like arvif. Due to this lock miss,
system get crashed in this function. Because if reading and writing
happen at the same time, there is a chance of crash or unexpected
behavior.

Remove the arvif lookup from ath12k_wmi_bcn_tmpl as it is already
present in the previous caller of the funciton.

Patch-dependency: 691-0004-ath12k-add-critical-update-support-during-set-beacon.patch

Signed-off-by: Balamurugan Ramar <quic_bramar@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/mac.c                        | 3 +--
 drivers/net/wireless/ath/ath12k/wmi.c                        | 5 +++--
 drivers/net/wireless/ath/ath12k/wmi.h                        | 3 ++-
 3 files changed, 6 insertions(+), 5 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/mac.c b/drivers/net/wireless/ath/ath12k/mac.c
index c5a4284..184be6d 100644
--- a/drivers/net/wireless/ath/ath12k/mac.c
+++ b/drivers/net/wireless/ath/ath12k/mac.c
@@ -1889,7 +1889,6 @@ static int __ath12k_mac_setup_bcn_tmpl(struct ath12k_link_vif *arvif,
 				       struct ieee80211_mutable_offsets offs,
 				       int ema_idx, int ema_cnt)
 {
-	struct ath12k *ar = arvif->ar;
 	u32 ema_param = 0;
 
 	if (ema_cnt) {
@@ -1901,7 +1900,7 @@ static int __ath12k_mac_setup_bcn_tmpl(struct ath12k_link_vif *arvif,
 			      WMI_BEACON_EMA_PARAM_LAST_TMPL_SHIFT);
 	}
 
-	return ath12k_wmi_bcn_tmpl(ar, arvif->vdev_id, &offs, bcn, ema_param);
+	return ath12k_wmi_bcn_tmpl(arvif, &offs, bcn, ema_param);
 }
 
 static int ath12k_mac_setup_bcn_tmpl_ema(struct ath12k_link_vif *arvif)
diff --git a/drivers/net/wireless/ath/ath12k/wmi.c b/drivers/net/wireless/ath/ath12k/wmi.c
index d987429..c44ef46 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.c
+++ b/drivers/net/wireless/ath/ath12k/wmi.c
@@ -3103,10 +3103,12 @@ ath12k_wmi_append_critical_update_params(struct ath12k *ar, u32 vdev_id,
 	return ptr;
 }
 
-int ath12k_wmi_bcn_tmpl(struct ath12k *ar, u32 vdev_id,
+int ath12k_wmi_bcn_tmpl(struct ath12k_link_vif *arvif,
 			struct ieee80211_mutable_offsets *offs,
 			struct sk_buff *bcn, u32 ema_params)
 {
+	struct ath12k *ar = arvif->ar;
+	u32 vdev_id = arvif->vdev_id;
 	struct ath12k_wmi_pdev *wmi = ar->wmi;
 	struct wmi_bcn_tmpl_cmd *cmd;
 	struct ath12k_wmi_bcn_prb_info_params *bcn_prb_info;
@@ -3115,7 +3117,6 @@ int ath12k_wmi_bcn_tmpl(struct ath12k *ar, u32 vdev_id,
 	void *ptr;
 	int ret, len;
 	size_t aligned_len = roundup(bcn->len, 4);
-	struct ath12k_link_vif *arvif = ath12k_mac_get_arvif(ar, vdev_id);
 	struct wmi_critical_update_arg cu_arg = {
 						 .num_ml_params = 0,
 						 .ml_params = NULL,
diff --git a/drivers/net/wireless/ath/ath12k/wmi.h b/drivers/net/wireless/ath/ath12k/wmi.h
index d69a4f7..0cc02ce 100644
--- a/drivers/net/wireless/ath/ath12k/wmi.h
+++ b/drivers/net/wireless/ath/ath12k/wmi.h
@@ -28,6 +28,7 @@ struct ath12k_fw_stats;
 struct ath12k_reg_tpc_power_info;
 struct ath12k_sawf_svc_params;
 struct ath12k_latency_params;
+struct ath12k_link_vif;
 
 #define PSOC_HOST_MAX_NUM_SS (8)
 
@@ -8732,7 +8733,7 @@ int ath12k_wmi_mgmt_send(struct ath12k *ar, u32 vdev_id, u32 buf_id,
 			 bool tx_params_valid);
 int ath12k_wmi_offchan_mgmt_send(struct ath12k *ar, u32 vdev_id, u32 buf_id,
 				 struct sk_buff *frame);
-int ath12k_wmi_bcn_tmpl(struct ath12k *ar, u32 vdev_id,
+int ath12k_wmi_bcn_tmpl(struct ath12k_link_vif *arvif,
 			struct ieee80211_mutable_offsets *offs,
 			struct sk_buff *bcn, u32 ema_param);
 int ath12k_wmi_vdev_down(struct ath12k *ar, u8 vdev_id);
-- 
2.34.1

